File: backend/app/schemas/streaming.py
"""schemas/streaming.py -- message contracts for the /ws/stream WebSocket.

The protocol is deliberately tiny JSON messages, both directions, tagged
by a "type" field:

Client -> Server:
    {"type": "frame", "features": [258 floats], "timestamp_ms": 173...}
    {"type": "reset"}

Server -> Client:
    {"type": "ready", ...StreamReadyMessage}     -- once, right after connect
    {"type": "update", ...StreamUpdateMessage}   -- after each inference tick
    {"type": "error", "message": "..."}          -- malformed frame, doesn't close the connection
"""

from __future__ import annotations

from pydantic import BaseModel, Field


class LandmarkFrameMessage(BaseModel):
    type: str = Field("frame", frozen=True)
    features: list[float]
    timestamp_ms: int | None = None


class ResetMessage(BaseModel):
    type: str = Field("reset", frozen=True)


class StreamReadyMessage(BaseModel):
    type: str = Field("ready", frozen=True)
    model_ready: bool
    is_synthetic_placeholder: bool
    feature_dim: int
    vocab_size: int
    window_frames: int


class WordPrediction(BaseModel):
    word: str
    confidence: float


class StreamUpdateMessage(BaseModel):
    type: str = Field("update", frozen=True)
    transcript: str
    words: list[WordPrediction]
    buffer_frames: int
    latency_ms: float
    fps: float


class StreamErrorMessage(BaseModel):
    type: str = Field("error", frozen=True)
    message: str

File: frontend/types/streaming.ts
/** Mirrors backend/app/schemas/streaming.py's WordPrediction. */
export interface WordPrediction {
  word: string;
  confidence: number;
}

/** Sent once by the server right after the WebSocket connects.
 * Mirrors backend/app/schemas/streaming.py's StreamReadyMessage. */
export interface StreamReadyMessage {
  type: "ready";
  model_ready: boolean;
  is_synthetic_placeholder: boolean;
  feature_dim: number;
  vocab_size: number;
  window_frames: number;
}

/** Sent after each inference tick. Mirrors StreamUpdateMessage. */
export interface StreamUpdateMessage {
  type: "update";
  transcript: string;
  words: WordPrediction[];
  buffer_frames: number;
  latency_ms: number;
  fps: number;
}

/** Mirrors StreamErrorMessage -- sent for a malformed individual frame;
 * does NOT close the connection. */
export interface StreamErrorMessage {
  type: "error";
  message: string;
}

export interface StreamPongMessage {
  type: "pong";
}

export type StreamServerMessage =
  | StreamReadyMessage
  | StreamUpdateMessage
  | StreamErrorMessage
  | StreamPongMessage;

/** Client -> server messages. Mirrors LandmarkFrameMessage / ResetMessage. */
export interface StreamFrameMessage {
  type: "frame";
  features: number[];
  timestamp_ms?: number;
}

export interface StreamResetMessage {
  type: "reset";
}

export interface StreamPingMessage {
  type: "ping";
}

export type StreamClientMessage = StreamFrameMessage | StreamResetMessage | StreamPingMessage;

/** One completed utterance archived into the on-screen conversation
 * history when the user resets/starts a new one -- a purely
 * frontend-side concept, no backend equivalent (the backend only ever
 * knows about "the current buffer's transcript"). */
export interface ConversationEntry {
  id: string;
  transcript: string;
  words: WordPrediction[];
  endedAt: number; // Date.now() timestamp
}
